aws-android-sdk-core/src/main/java/com/amazonaws/internal/ReturningRunnable.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019-2019 Amazon.com, Inc. or its affiliates.
+ * All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.amazonaws.internal;
+
+import com.amazonaws.async.Callback;
+
+/**
+ * Internal class to allow easy surfacing of synchronous and asynchronous operations.
+ *
+ * @param <R> the type of the result/return
+ */
+public abstract class ReturningRunnable<R> {
+
+    private final String operationDescription;
+
+    public ReturningRunnable() {
+        this.operationDescription = null;
+    }
+
+    public ReturningRunnable(final String operationDescription) {
+        this.operationDescription = operationDescription;
+    }
+
+    public abstract R run() throws Exception;
+
+    /**
+     * Runs the code inside {@link #run()} synchronously on the same thread this method is called.
+     *
+     * @return the result from the {@link #run()} method
+     * @throws Exception error from the operation
+     */
+    public R await() throws Exception {
+        return this.run();
+    }
+
+    /**
+     * Runs the code inside {@link #run()} on a new Thread and returns immediately.
```

Comment:
on a new thread `asynchronously` and returns immediately.  The `callback` will be invoked when the operation is complete.